Fix an init-deadlock in OPENSSL_INIT_LOAD_CRYPTO_STRINGS

bernd-edlinger opened the pull request:
This can be reproduced with my error injection patch.

The test vector has been validated on the master branch:

$ ERROR_INJECT=1692279870 ../util/shlib_wrap.sh ./asn1parse-test ./corpora/asn1parse/027f6e82ba01d9db9a9167b83e56cc9f2c602550 ERROR_INJECT=1692279870
    #0 0x7f280b42fef8 in __sanitizer_print_stack_trace ../../../../src/libsanitizer/asan/asan_stack.cpp:86
    #1 0x5610a3f396b4 in my_malloc fuzz/test-corpus.c:114
    #2 0x7f280a2eb94c in CRYPTO_malloc crypto/mem.c:177
    #3 0x7f280a2dafdb in OPENSSL_LH_insert crypto/lhash/lhash.c:114
    #4 0x7f280a1c87fe in err_load_strings crypto/err/err.c:264
    #5 0x7f280a1c87fe in err_load_strings crypto/err/err.c:259
    #6 0x7f280a1c87fe in ERR_load_strings_const crypto/err/err.c:301
    #7 0x7f280a6f513b in ossl_err_load_PROV_strings providers/common/provider_err.c:233
    #8 0x7f280a1cf015 in ossl_err_load_crypto_strings crypto/err/err_all.c:109
    #9 0x7f280a2e9b8c in ossl_init_load_crypto_strings crypto/init.c:190
    #10 0x7f280a2e9b8c in ossl_init_load_crypto_strings_ossl_ crypto/init.c:181
    #11 0x7f2808cfbf67  (/lib/x86_64-linux-gnu/libc.so.6+0x99f67)
    #12 0x7f280a32301e in CRYPTO_THREAD_run_once crypto/threads_pthread.c:154
    #13 0x7f280a2ea1da in OPENSSL_init_crypto crypto/init.c:553
    #14 0x5610a3f38e2f in FuzzerInitialize fuzz/asn1parse.c:29
    #15 0x5610a3f38783 in main fuzz/test-corpus.c:194
    #16 0x7f2808c8bd8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f)
    #17 0x7f2808c8be3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f)
    #18 0x5610a3f38d34 in _start (/home/runner/work/openssl/openssl/fuzz/asn1parse-test+0x3d34)

AddressSanitizer:DEADLYSIGNAL
================================================================= ==27629==ERROR: AddressSanitizer: ABRT on unknown address 0x03e900006e23 (pc 0x7f2808cfbef8 bp 0x7f280b36afe0 sp 0x7ffd545b2460 T0)
    #0 0x7f2808cfbef8  (/lib/x86_64-linux-gnu/libc.so.6+0x99ef8)
    #1 0x7f280a32301e in CRYPTO_THREAD_run_once crypto/threads_pthread.c:154
    #2 0x7f280a2ea1da in OPENSSL_init_crypto crypto/init.c:553
    #3 0x7f280a1c935e in ossl_err_get_state_int crypto/err/err.c:705
    #4 0x7f280a1cf1f9 in ERR_new crypto/err/err_blocks.c:20
    #5 0x7f280a2eb9ac in CRYPTO_malloc crypto/mem.c:205
    #6 0x7f280a2dafdb in OPENSSL_LH_insert crypto/lhash/lhash.c:114
    #7 0x7f280a1c87fe in err_load_strings crypto/err/err.c:264
    #8 0x7f280a1c87fe in err_load_strings crypto/err/err.c:259
    #9 0x7f280a1c87fe in ERR_load_strings_const crypto/err/err.c:301
    #10 0x7f280a6f513b in ossl_err_load_PROV_strings providers/common/provider_err.c:233
    #11 0x7f280a1cf015 in ossl_err_load_crypto_strings crypto/err/err_all.c:109
    #12 0x7f280a2e9b8c in ossl_init_load_crypto_strings crypto/init.c:190
    #13 0x7f280a2e9b8c in ossl_init_load_crypto_strings_ossl_ crypto/init.c:181
    #14 0x7f2808cfbf67  (/lib/x86_64-linux-gnu/libc.so.6+0x99f67)
    #15 0x7f280a32301e in CRYPTO_THREAD_run_once crypto/threads_pthread.c:154
    #16 0x7f280a2ea1da in OPENSSL_init_crypto crypto/init.c:553
    #17 0x5610a3f38e2f in FuzzerInitialize fuzz/asn1parse.c:29
    #18 0x5610a3f38783 in main fuzz/test-corpus.c:194
    #19 0x7f2808c8bd8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f)
    #20 0x7f2808c8be3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f)
    #21 0x5610a3f38d34 in _start (/home/runner/work/openssl/openssl/fuzz/asn1parse-test+0x3d34)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: ABRT (/lib/x86_64-linux-gnu/libc.so.6+0x99ef8) ==27629==ABORTING

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

slontis commented:
It might be good to include the offending OPENSSL_init_crypto() call in the commit message?

t8m commented on crypto/err/err.c:
```diff
@@ -454,6 +454,9 @@ static unsigned long get_error_values(ERR_GET_ACTION g,
     if (es == NULL)
         return 0;
 
+    if (!OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL))
+        return 0;
+
```
Should it be moved to ossl_err_string_int() instead?

t8m commented:
Also the pyca external test failure seems relevant. Unfortunately it won't be trivial to fix I am afraid.

bernd-edlinger commented:
@t8m I tried to follow your suggestion, but since `ossl_err_string_int`
is a void return function there is no way to return an error code when
`OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL)` fails.
why do you prefer that function over `get_error_values` ?

bernd-edlinger commented:
No, this does not work at all, and caused:
```
65-test_cmp_ctx.t .. 
CMP error: reason(102)
    # total=11 len=11 msg='reason(100)'
    # 
    # total=45 len=34 msg='reason(103):data1 : data2
    # new line'
    # 
    # ERROR: (int) 'msg_total_size == expected_size' failed @ test/cmp_ctx_test.c:185
    # [45] compared to [48]
    # total=3605 len=3605 msg='reason(100):This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 61 bytes loooooooooooooooooooooooooooooong string.
    # <SEP>This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 61 bytes loooooooooooooooooooooooooooooong string.
    # <SEP>This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 61 bytes loooooooooooooooooooooooooooooong string.
    # <SEP>This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 61 bytes loooooooooooooooooooooooooooooong string.
    # <SEP>This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 61 bytes loooooooooooooooooooooooooooooong string.
    # <SEP>This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 61 bytes loooooooooooooooooooooooooooooong string.
    # <SEP>This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 61 bytes loooooooooooooooooooooooooooooong string.
    # '
    # 
    # total=4126 len=521 msg='reason(100):This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 64 bytes looooooooooooooooooooooooooooooooong string.
    # This is a 61 bytes loooooooooooooooooooooooooooooong string.
    # '
    # 
    # ERROR: (int) 'msg_total_size == expected_size' failed @ test/cmp_ctx_test.c:198
    # [4126] compared to [4128]
    # OPENSSL_TEST_RAND_ORDER=1691519185
    not ok 6 - test_CTX_print_errors
# ------------------------------------------------------------------------------
../../util/wrap.pl ../../test/cmp_ctx_test => 1
not ok 1
65-test_cmp_ctx.t .. 1/? -------------------------------------------------------
#   Failed test at test/recipes/65-test_cmp_ctx.t line 24.
65-test_cmp_ctx.t .. Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
```

t8m commented on crypto/err/err.c:
```diff
@@ -589,9 +589,10 @@ const char *ERR_lib_error_string(unsigned long e)
     unsigned long l;
 
-    if (!RUN_ONCE(&err_string_init, do_err_strings_init)) {
+    if (!OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL)
```
Perhaps you need to also add it to ERR_reason_error_string() ?

t8m commented:
The ci failure is quite strange now.

t8m reviewed (approved):
Could you please amend the commit message so it tells about the final fix and not the previous attempts? Also I do not think it needs all the backtraces but it is up to you how detailed you want it to be.

bernd-edlinger commented:
I updated the commit message once more: I added this statement describing how the fix works (and removed the accidental stuff at the end):

    A recursive OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS) call
    may happen if an out-of-memory error happens at the first callstack,
    and the dead-lock happens at the second callstack, because ossl_err_get_state_int
    calls OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS) although that
    call is currently already executing.
    
    At least on posix system this causes the process to freeze at this
    point, and must be avoided whatever it takes.
    
    The fix is using err_shelve_state around the critical region, which
    makes ossl_err_get_state_int return early and not call the recursive
    OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS).

But I have still no Idea why 
147e8444d3a467b30f116ea507d7117c035f713e
made the test_shlibload fail...

tom-cosgrove-arm commented on crypto/init.c:
```diff
@@ -186,8 +186,15 @@ DEFINE_RUN_ONCE_STATIC(ossl_init_load_crypto_strings)
+        return 0;
+
     OSSL_TRACE(INIT, "ossl_err_load_crypto_strings()\n");
     ret = ossl_err_load_crypto_strings();
```
Can I check that we don't want to call `ossl_err_load_crypto_strings()` at all if either `OPENSSL_NO_ERR` or `OPENSSL_NO_AUTOERRINIT` are defined?

tom-cosgrove-arm reviewed (approved):
LGTM

bernd-edlinger commented:
Merged to master as 6b1a127.  Thanks!
